server/src/main/java/org/opensearch/rest/RequestLimitSettings.java

```diff
@@ -0,0 +1,174 @@
+ */
+public class RequestLimitSettings {
+
+    /**
+     * Enum to represent entity against which we need to perform limit checks.
+     */
+    public enum BlockEntity {
+        INDICES,
+        SHARDS
+    }
+
+    private volatile int catIndicesLimit;
+    private volatile int catShardsLimit;
+    private volatile int catSegmentsLimit;
+
+    /**
+     * Setting to enable circuit breaker on {@link RestIndicesAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_INDICES_LIMIT_SETTING = Setting.intSetting(
+        "cat.indices.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable circuit breaker on {@link RestShardsAction}. The limit will be applied on number of shards.
+     */
+    public static final Setting<Integer> CAT_SHARDS_LIMIT_SETTING = Setting.intSetting(
+        "cat.shards.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Setting to enable circuit breaker on {@link RestSegmentsAction}. The limit will be applied on number of indices.
+     */
+    public static final Setting<Integer> CAT_SEGMENTS_LIMIT_SETTING = Setting.intSetting(
+        "cat.segments.limit",
+        -1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public RequestLimitSettings(ClusterSettings clusterSettings, Settings settings) {
+        setCatShardsLimit(CAT_SHARDS_LIMIT_SETTING.get(settings));
+        setCatIndicesLimit(CAT_INDICES_LIMIT_SETTING.get(settings));
+        setCatSegmentsLimit(CAT_SEGMENTS_LIMIT_SETTING.get(settings));
+
+        clusterSettings.addSettingsUpdateConsumer(CAT_SHARDS_LIMIT_SETTING, this::setCatShardsLimit);
+        clusterSettings.addSettingsUpdateConsumer(CAT_INDICES_LIMIT_SETTING, this::setCatIndicesLimit);
+        clusterSettings.addSettingsUpdateConsumer(CAT_SEGMENTS_LIMIT_SETTING, this::setCatSegmentsLimit);
+    }
+
+    /**
+     * Method to check if the circuit breaker limit has reached for an action.
+     * The limits are controlled via dynamic settings.
+     *
+     * @param metadata     {@link Metadata}
+     * @param blockEntity  {@link BlockEntity}
+     * @param limit        Integer limit on block entity
+     * @return True/False
+     */
+    public static boolean isCircuitLimitBreached(final Metadata metadata, final BlockEntity blockEntity, final int limit) {
+        if (Objects.isNull(metadata)) return false;
+        if (limit <= 0) return false;
+        switch (blockEntity) {
+            case INDICES:
+                int indicesCount = chainWalk(() -> metadata.getIndices().size(), 0);
+                if (indicesCount > limit) return true;
+                break;
+            case SHARDS:
+                int shardsCount = metadata.getTotalNumberOfShards();
+                if (shardsCount > limit) return true;
+                break;
+        }
```

Comment:
The metadata is not available in cat shards scenario so we had to use routing table.

I have remove case to use value from metadata since its not required for now.